pkg/agent/bgp/gobgp/gobgp.go

```diff
@@ -0,0 +1,299 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package gobgp
+
+import (
+	"context"
+	"fmt"
+	"net/netip"
+	"time"
+
+	gobgpapi "github.com/osrg/gobgp/v3/api"
+	"github.com/osrg/gobgp/v3/pkg/server"
+	"google.golang.org/protobuf/types/known/anypb"
+	"k8s.io/utils/net"
+
+	"antrea.io/antrea/pkg/agent/bgp"
+)
+
+const (
+	ipv4AllZero = "0.0.0.0"
+	ipv6AllZero = "::"
+)
+
+type Server struct {
+	server       *server.BgpServer
+	globalConfig *gobgpapi.Global
+}
+
+func NewGoBGPServer(globalConfig *bgp.GlobalConfig) bgp.Interface {
+	return &Server{
+		server: server.NewBgpServer(server.LoggerOption(newGoBGPLogger())),
+		globalConfig: &gobgpapi.Global{
+			Asn:        globalConfig.ASN,
+			RouterId:   globalConfig.RouterID,
+			ListenPort: globalConfig.ListenPort,
+		},
+	}
+}
+
+func (s *Server) Start(ctx context.Context) error {
+	go s.server.Serve()
+	if err := s.server.StartBgp(ctx, &gobgpapi.StartBgpRequest{Global: s.globalConfig}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) Stop(ctx context.Context) error {
+	if err := s.server.StopBgp(ctx, &gobgpapi.StopBgpRequest{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) AddPeer(ctx context.Context, peer bgp.PeerConfig) error {
+	request := &gobgpapi.AddPeerRequest{Peer: toGoBGPPeer(peer)}
+	if err := s.server.AddPeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) UpdatePeer(ctx context.Context, peer bgp.PeerConfig) error {
+	request := &gobgpapi.UpdatePeerRequest{Peer: toGoBGPPeer(peer)}
+	if _, err := s.server.UpdatePeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) RemovePeer(ctx context.Context, peer bgp.PeerConfig) error {
+	request := &gobgpapi.DeletePeerRequest{Address: peer.Address}
+	if err := s.server.DeletePeer(ctx, request); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *Server) GetPeers(ctx context.Context) ([]bgp.PeerStatus, error) {
+	var peerStatuses []bgp.PeerStatus
+
+	fn := func(peer *gobgpapi.Peer) {
+		if peer == nil {
+			return
+		}
+		peerStatus := bgp.PeerStatus{}
+
+		// According to the gobgp code, all pointer fields in `peer *gobgpapi.Peer` used below should be set and non-nil
+		// when peers are listed. It's safe and harmless to keep the nil checks.
+		if transport := peer.GetTransport(); transport != nil {
+			peerStatus.Port = int32(transport.GetRemotePort())
+		}
+		if conf := peer.GetConf(); conf != nil {
+			peerStatus.Address = conf.GetNeighborAddress()
+			peerStatus.ASN = int32(conf.GetPeerAsn())
+		}
+		if ebgpMultiHop := peer.GetEbgpMultihop(); ebgpMultiHop != nil {
+			peerStatus.MultihopTTL = int32(ebgpMultiHop.GetMultihopTtl())
+		} else {
+			peerStatus.MultihopTTL = bgp.DefaultBGPMultihopTTL
+		}
```

Comment:
Remove the `else` case. I agree that keeping the `nil` checks is good, but I would not include code to default `peerStatus.MultihopTTL` and `peerStatus.GracefulRestartTimeSeconds` if the corresponding fields are `nil` (given that `nil` indicates a bug somewhere, and the `nil` checks constitute defensive programming).

Reply:
> if the corresponding fields are nil (given that nil indicates a bug somewhere, and the nil checks constitute defensive programming).

Thanks for explaining this. My understanding is that we should not handle the bug potentially caused by gobgp? Is my understanding correct?